samples/star/src/main/kotlin/com/slack/circuit/star/data/Petfinder.kt

```diff
@@ -171,53 +172,47 @@ data class AuthenticationData(
  * A hypothetical token manager that stores an auth token. Just in-memory and not thread-safe for
  * now.
  */
-class TokenManager(private val api: PetfinderAuthApi) {
-  // TODO store this somewhere
-  private var expiration: Instant = Instant.EPOCH
-  private var tokenType: String? = null
-  private var token: String? = null
-
-  /** Returns a token if */
-  fun authenticate(request: Request): Request {
+class TokenManager(private val api: PetfinderAuthApi, private val tokenStorage: TokenStorage) {
+  private val mutex = Mutex()
+
+  /** Authenticates a [request]. */
+  suspend fun authenticate(request: Request): Request {
     return authenticate(request, false)
   }
 
-  private fun authenticate(request: Request, isAfterRefresh: Boolean): Request {
+  private suspend fun authenticate(request: Request, isAfterRefresh: Boolean): Request {
     println("INFO: Authenticating request ${request.url}")
     val newBuilder = request.newBuilder()
-    if (Instant.now().isAfter(expiration)) {
+    val (tokenType, expiration, token) =
+      tokenStorage.getAuthData()
+        ?: run {
+          runBlocking { refreshToken() }
+          return authenticate(request, isAfterRefresh)
+        }
+    if (Instant.now().isAfter(Instant.ofEpochMilli(expiration))) {
       check(!isAfterRefresh)
       runBlocking { refreshToken() }
       return authenticate(request, isAfterRefresh)
     } else {
-      val localTokenType = tokenType
-      val localToken = token
-      if (localToken != null && localTokenType != null) {
-        newBuilder.addHeader("Authorization", "$localTokenType $localToken")
-      } else {
-        check(!isAfterRefresh)
-        runBlocking { refreshToken() }
-        return authenticate(request, isAfterRefresh)
-      }
+      newBuilder.addHeader("Authorization", "$tokenType $token")
     }
 
     return newBuilder.build()
   }
 
-  private suspend fun refreshToken() {
-    println("INFO: Refreshing token")
+  private suspend fun refreshToken() =
+    mutex.withLock {
```

Comment:
opportunistic, solution to the threadsafe/de-dupe comment below. It doesn't de-dupe, but at least avoids the races